Lifecycle Methods and Event Alternative to `sessionsId` (#285)
ChainAgnostic/CAIPs: BCP/Session Lifecycle Management.md

Additionally, it allows for session management without mandatory [sessionIds][CAIP-171], offering more flexibility in handling sessions in single-session contexts.

## Abstract

This proposal aims to extend the CAIP-25 standard by defining new JSON-RPC methods for managing the lifecycle of authorizations within a session. These methods allow dapps and wallets to dynamically adjust authorizations, providing more granular control and better user experience. Additionally, it allows for session management without mandatory sessionIds, offering more flexibility in handling sessions.

## Motivation

The motivation behind this proposal is to enhance the flexibility of CAIP-25 by enabling management of session authorizations without sessionIds which don't map well to extension based wallet's dapp connections and could therefore add constraints and burdens to existing flows. The proposed methods provide an intuitive way to add, revoke, and retrieve authorizations within an existing session, simplifying the management of session lifecycles.

## Lifecycle Equivalence Chart

> ||feature|CAIP-25 now w/sessionId|CAIP-285 w/o sessionId|

Reviewer: ```suggestion
||feature|CAIP-25 now w/ sessionId|CAIP-285 w/o sessionId|
```

Author: fixed here: https://github.com/ChainAgnostic/CAIPs/pull/285/commits/03e422eb7078ae2a1bc618c383133fcaa767b662

|---|---|---|---|

> |1|dapp initialize (replaces session if already exist)|call `wallet_createSession` w/no sessionId |call `wallet_createSession` w/no sessionId|

Reviewer: ```suggestion
|1|dapp initialize (replaces session if already exist)|call `wallet_createSession` w/o sessionId |call `wallet_createSession` w/o sessionId|
```

Author: fixed here: https://github.com/ChainAgnostic/CAIPs/pull/285/commits/03e422eb7078ae2a1bc618c383133fcaa767b662

|2|wallet re-initialize|return `wallet_createSession` w/new sessionId **next time called**|n/a (not needed because [`wallet_sessionChanged`][CAIP-308] notification can be sent, and wallet_getSession can be used to confirm everything is good)|
|3|dapp get current session|n/a (should persist)|[`wallet_getSession`][CAIP-307]  w/o sessionId|
|4|dapp confirm current session|call `wallet_createSession` w/sessionId and same properties OR `wallet_getSession` w/sessionId|`wallet_getSession` w/o sessionId|

> |5|dapp revoke|call `wallet_createSession` w/no sessionId and no scopes OR [`wallet_revokeSession`][CAIP-285] w/sessionId |`wallet_revokeSession`  w/o sessionId|

Reviewer: ```suggestion
|5|dapp revoke|call `wallet_createSession` w/o sessionId and no scopes OR [`wallet_revokeSession`][CAIP-285] w/sessionId |`wallet_revokeSession`  w/o sessionId|
```

Author: fixed here: https://github.com/ChainAgnostic/CAIPs/pull/285/commits/03e422eb7078ae2a1bc618c383133fcaa767b662

> |6|wallet revoke|return `wallet_createSession` w/new sessionId and no scopes **next time called** or `wallet_sessionChanged` w/ sessionId |`wallet_sessionChanged`  w/no scopes|

Reviewer: ```suggestion
|6|wallet revoke|return `wallet_createSession` w/new sessionId and no scopes **next time called** or `wallet_sessionChanged` w/ sessionId |`wallet_sessionChanged`  w/o scopes|
```

Author: fixed here: https://github.com/ChainAgnostic/CAIPs/pull/285/commits/03e422eb7078ae2a1bc618c383133fcaa767b662

> |7|dapp update session|call `wallet_createSession` w/existing sessionId and new scopes|call `wallet_createSession` w/no sessionId|

Reviewer: ```suggestion
|7|dapp update session|call `wallet_createSession` w/existing sessionId and new scopes|call `wallet_createSession` w/o sessionId|
```

Author: fixed here: https://github.com/ChainAgnostic/CAIPs/pull/285/commits/03e422eb7078ae2a1bc618c383133fcaa767b662

|8|wallet update session|return `wallet_createSession` w/new sessionId and no scopes **next time called** OR `wallet_sessionChanged` w/existing sessionId|`wallet_sessionChanged` w/o sessionId|

## Lifecycle diagrams

### Visualizing the lifecycle of a session without a `sessionId`

```mermaid
sequenceDiagram
  rect rgb(255, 255, 255)
  actor Caller
  participant Wallet
  participant WalletDataStore as Data Store (Wallet)
  rect rgb(255, 245, 245)
  note right of Caller: Session Creation
  Caller->>Wallet: wallet_createSession
  Wallet->>WalletDataStore: Persist session data
  Wallet-->>Caller: {"sessionScopes": {...}}
  end
  rect rgb(245, 255, 245)
  note right of Caller: Update Session
  Caller->>Wallet: wallet_createSession (update auth)
  Wallet->>WalletDataStore: Update session data
  Wallet-->>Caller: {"sessionScopes": {updatedScopes...}}
  end
  rect rgb(245, 245, 255)
  note right of Caller: Connection Interrupted w/ Wallet Side Session Modification
  Caller-->Wallet: Connection Interrupted
  Wallet->>WalletDataStore: User initiated session change

>   Wallet->>Caller: wallet_sessionChanged

Reviewer: this notification can't be sent since connection is interrupted

Author: Yeah I think the idea here is that it attempts to send it but its not received. Will see if there is a better way to represent this in the diagram

  Caller-->Wallet: Connection Re-established
  end
  rect rgb(255, 245, 255)
  note right of Caller: Get Session
  Caller->>Wallet: wallet_getSession
  Wallet-->>Caller: {"sessionScopes": {...}}
  end
  rect rgb(255, 245, 215)
  note right of Caller: Revoke Session
  Caller->>Wallet: wallet_revokeSession
  Wallet->>WalletDataStore: Update session data

>   Wallet-->>Caller: {"result": "session revoked"}

Reviewer: ```suggestion
  Wallet-->>Caller: {"result": true}
```

Author: fixed here: https://github.com/ChainAgnostic/CAIPs/pull/285/commits/03e422eb7078ae2a1bc618c383133fcaa767b662

  end
  end
```

### Visualizing the lifecycle of a session **with** a `sessionId`

```mermaid
sequenceDiagram
  rect rgb(255, 255, 255)
  participant CallerDataStore as Data Store (Caller)
  actor Caller
  participant Wallet
  participant WalletDataStore as Data Store (Wallet)
  rect rgb(255, 245, 245)
  note right of Caller: Session Creation
  Caller->>Wallet: wallet_createSession
  Wallet->>WalletDataStore: Persist session data
  Wallet-->>Caller: {"sessionId": "0xdeadbeef", "sessionScopes": {...}}
  Caller->>CallerDataStore: Persist session data
  end
  rect rgb(245, 255, 245)
  note right of Caller: Update Session
  Caller->>Wallet: wallet_createSession (sessionId: 0xdeadbeef, {updatedScopes...})
  Wallet->>WalletDataStore: Update session data
  Wallet-->>Caller: {"sessionId": "0xdeadbeef", "sessionScopes": {(updated)sessionScopes...}}

>   end

Reviewer: should a persist session data for the caller go here?

Author: fixed here: https://github.com/ChainAgnostic/CAIPs/pull/285/commits/03e422eb7078ae2a1bc618c383133fcaa767b662

  rect rgb(245, 245, 255)
  note right of Caller: User Initiated Session Change
  Wallet->>WalletDataStore: User initiated session change
  Wallet->>Caller: wallet_sessionChanged (sessionId: 0xdeadbeef)
  Caller->>CallerDataStore: Update session data
  end
  rect rgb(255, 245, 255)
  note right of Caller: Revoke Session
  Caller->>Wallet: wallet_revokeSession (sessionId: 0xdeadbeef)
  Wallet->>WalletDataStore: Update session data
  Wallet-->>Caller: {"result": "true"} (session is revoked)
  Caller->>CallerDataStore: Update session data
  end
  end
```
## Privacy Considerations

> MThe introduction of this lifecycle method must ensure that only authorized parties can retrieve the authorizations of a session. Proper authentication and authorization mechanisms must be in place to prevent unauthorized access or modifications.

Reviewer: ```suggestion
The introduction of this lifecycle method must ensure that only authorized parties can retrieve the authorizations of a session. Proper authentication and authorization mechanisms must be in place to prevent unauthorized access or modifications.
```

Author: fixed here: https://github.com/ChainAgnostic/CAIPs/pull/285/commits/03e422eb7078ae2a1bc618c383133fcaa767b662

---

> title: JSON-RPC Provider Session Lifecycle Management with CAIP-25 Sessions BCP

Reviewer 2: ```suggestion
caip: 316
title: JSON-RPC Provider Session Lifecycle Management with CAIP-25 Sessions BCP
```

and i'd say just rename the file `caip-316` and move it into the CAIPs folder!

author: [Alex Donesky] (@adonesky1)

> discussions-to: TBD

Reviewer 2: ```suggestion
discussions-to: https://github.com/ChainAgnostic/CAIPs/pull/285/files
```

status: Draft

> type: Standard

Reviewer 2: ```suggestion
type: Informational
```

created: 2024-06-07

> requires: 25, 217, 285, 307, 308

Reviewer 2: ```suggestion
requires: 25, 217, 285, 311, 312
```

---

## Simple Summary

> This backwards-compatible extension of the [CAIP-25][] standard defines new JSON-RPC methods for managing the lifecycle of authorizations within a session.

Reviewer 2: ```suggestion
This overview compares new and old forms of [CAIP-25][] connection flow to ensure safe and well-informed implementation of either or both.
[CAIP-311][] and [CAIP-312][] extended [CAIP-25][] by defining new JSON-RPC methods for managing the lifecycle of authorizations within a session.
```

These methods allow dapps and wallets to dynamically adjust authorizations, providing more granular control and better user experience.
Additionally, it allows for session management without mandatory [sessionIds][CAIP-171], offering more flexibility in handling sessions in single-session contexts.

## Abstract

> This proposal aims to extend the CAIP-25 standard by defining new JSON-RPC methods for managing the lifecycle of authorizations within a session. These methods allow dapps and wallets to dynamically adjust authorizations, providing more granular control and better user experience. Additionally, it allows for session management without mandatory sessionIds, offering more flexibility in handling sessions.

Reviewer 2: By loosening the earlier requirement that both caller and wallet maintain a session identifier to track the session across potentially shared or insecure transports, single-session transport options can take advantage of direct RPC calls to do more explicit session-management.
The equivalence across different transports can be counter-intuitive, so the following diagrams and examples are provided for apples-to-apples comparisons.

> This proposal aims to extend the CAIP-25 standard by defining new JSON-RPC methods for managing the lifecycle of authorizations within a session. These methods allow dapps and wallets to dynamically adjust authorizations, providing more granular control and better user experience. Additionally, it allows for session management without mandatory sessionIds, offering more flexibility in handling sessions.

Reviewer 2: ```suggestion
By loosening the earlier requirement that both caller and wallet maintain a session identifier to track the session across potentially shared or insecure transports, single-session transport options can take advantage of direct RPC calls to do more explicit session-management.
The equivalence across different transports can be counter-intuitive, so the following diagrams and examples are provided for apples-to-apples comparisons.
```

## Motivation

> The motivation behind this proposal is to enhance the flexibility of CAIP-25 by enabling management of session authorizations without sessionIds which don't map well to extension based wallet's dapp connections and could therefore add constraints and burdens to existing flows. The proposed methods provide an intuitive way to add, revoke, and retrieve authorizations within an existing session, simplifying the management of session lifecycles.

Reviewer 2: ```suggestion
The equivalence across different transports can be counter-intuitive, so the following diagrams and examples are provided for apples-to-apples comparisons.
```

```mermaid
sequenceDiagram
  rect rgb(255, 255, 255)
  participant CallerDataStore as Data Store (Caller)
  actor Caller
  participant Wallet
  participant WalletDataStore as Data Store (Wallet)
  rect rgb(255, 245, 245)
  note right of Caller: Session Creation
  Caller->>Wallet: wallet_createSession
  Wallet->>WalletDataStore: Persist session data
  Wallet-->>Caller: {"sessionId": "0xdeadbeef", "sessionScopes": {...}}
  Caller->>CallerDataStore: Persist session data
  end
  rect rgb(245, 255, 245)
  note right of Caller: Update Session
  Caller->>Wallet: wallet_createSession (sessionId: 0xdeadbeef, {updatedScopes...})
  Wallet->>WalletDataStore: Update session data
  Wallet-->>Caller: {"sessionId": "0xdeadbeef", "sessionScopes": {(updated)sessionScopes...}}
  Caller->>CallerDataStore: Persist session data
  end
  rect rgb(245, 245, 255)
  note right of Caller: User Initiated Session Change
  Wallet->>WalletDataStore: User initiated session change
  Wallet->>Caller: wallet_sessionChanged (sessionId: 0xdeadbeef)
  Caller->>CallerDataStore: Update session data
  end
  rect rgb(255, 245, 255)
  note right of Caller: Revoke Session

>   Caller->>Wallet: wallet_revokeSession (sessionId: 0xdeadbeef)

Author: I suppose this should be 
```suggestion
  Caller->>Wallet: wallet_createSession (no sessionId, no scopes)
```
@bumblefudge its not clear what a full revoke response should be?

Reviewer 2: Actually both work-- if wallet SUPPORTS `wallet_revokeSession` (which the caller would know if it was authorized to call it in the scopeObject !), they could call it with the optional sessionId, or if the dapp doesn't know for a fact that i CAN explicitly use that method, any `wallet_createSession` with a new sessionId is expected to kill the previous sessionId-having session.  

(if I remember correctly, we didn't make it a MUST on the wallet side to kill multiple sessionId`s per dapp because it out would be hard to test/guarantee/prove.)

Reviewer 2: ```suggestion
  Caller->>Wallet: wallet_createSession (sessionId: 0xnewbeef, no scopes)
  Wallet->>WalletDataStore: Create new, empty session 0xnewbeef ; clear all older sessions w/same dapp
  Wallet-->>Caller: {"result": "true"} (session is revoked)
  Caller->>CallerDataStore: Clear session data
  end
  rect rgb(255, 245, 255)
  note right of Caller: Revoke Session
  Caller->>Wallet: wallet_revokeSession (sessionId: 0xdeadbeef)
```

Reviewer 3: > if wallet SUPPORTS wallet_revokeSession (which the caller would know if it was authorized to call it in the scopeObject !)

Wait, do we want this to be a permissioned method? Seems like that these lifecycle methods should be non-permissioned?

Reviewer 2: Oh, I guess here we're in the overlap of feature discovery overlapping with feature permissioning! What I meant was, as written, getting a connection **without** sessionId back from a wallet means a dapp can **assume** support for `wallet_revokeSession`, but getting a connection **with** sessionId back means... they don't know if they have to do it the legacy way or if they also have the option to do it the explicit way.  so a wallet could explicitly authorize that in `sessionScopes` (even if not requested!), but otherwise, how will the dapp know?

Reviewer 3: But isn't `wallet_revokeSession` able to accept both with or without `sessionId`? If they get back the `sessionId`, that implicitly tells the dapp that the wallet is tracking that session with that id, so they should be calling it with that session id no? And then they can call it without `sessionId` if they don't get anything back?

>   Caller->>Wallet: wallet_revokeSession (sessionId: 0xdeadbeef)
>   Wallet->>WalletDataStore: Update session data
>   Wallet-->>Caller: {"result": "true"} (session is revoked)

Author: I suppose this should be 
```suggestion
  Caller->>Wallet: wallet_createSession (no sessionId, no scopes)
```
@bumblefudge its not clear what a full revoke response should be?

  Caller->>CallerDataStore: Update session data
  end
  end
```
## Privacy Considerations

The introduction of this lifecycle method must ensure that only authorized parties can retrieve the authorizations of a session. Proper authentication and authorization mechanisms must be in place to prevent unauthorized access or modifications.

> To achieve this, it is recommended to establish a connection over domain-bound or other 1:1 transports. Where applicable additional binding to a `sessionId` is recommended to ensure secure session management. This approach helps to create a secure communication channel that can effectively authenticate and authorize session-related requests, minimizing the risk of unauthorized access or session hijacking.

Reviewer 2: ```suggestion
To achieve this, it is recommended to establish a connection over domain-bound or other 1:1 transports. Where applicable, additional binding to a `sessionId` is recommended to ensure secure session management. This approach helps to create a secure communication channel that can effectively authenticate and authorize session-related requests, minimizing the risk of unauthorized access or session hijacking.
```
